llms/mlx_lm/models/minicpm.py

```diff
@@ -0,0 +1,203 @@
+
+
+class MiniCPMDecoderLayer(nn.Module):
+    def __init__(self, args: ModelArgs) -> None:
+        super().__init__()
+        self.args = args
+        self.hidden_size = args.hidden_size
+        self.num_hidden_layers = args.num_hidden_layers
+
+        self.self_attn = MiniCPMAttention(args)
+        self.mlp = MiniCPMMLP(args)
+        self.input_layernorm = MiniCPMRMSNorm(args.hidden_size, eps=args.rms_norm_eps)
+        self.post_attention_layernorm = MiniCPMRMSNorm(args.hidden_size, eps=args.rms_norm_eps)
+
+        self.scale_depth = args.scale_depth
+
+    def __call__(
+        self,
+        x: mx.array,
+        mask: Optional[mx.array] = None,
+        cache: Optional[Tuple[mx.array, mx.array]] = None,
+    ) -> mx.array:
+        r, cache = self.self_attn(self.input_layernorm(x), mask, cache)
+        h = x + r * (self.scale_depth / np.sqrt(self.num_hidden_layers))
+        h = self.mlp(self.post_attention_layernorm(h))
+        out = x + r * (self.scale_depth / np.sqrt(self.num_hidden_layers))
+        return out, cache
+
+
+class MiniCPMModel(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.args = args
+        self.vocab_size = args.vocab_size
+        assert self.vocab_size > 0
+
+        self.embed_tokens = nn.Embedding(args.vocab_size, args.hidden_size)
+        self.layers = [
+                MiniCPMDecoderLayer(args=args) for _ in range(args.num_hidden_layers)
+        ]
+        self.norm = MiniCPMRMSNorm(args.hidden_size, eps=args.rms_norm_eps)
+
+    def __call__(
+        self,
+        inputs: mx.array,
+        cache=None,
+    ):
+        h = self.embed_tokens(inputs)
+
+        mask = None
+        if h.shape[1] > 1:
+            mask = nn.MultiHeadAttention.create_additive_causal_mask(h.shape[1])
+            mask = mask.astype(h.dtype)
+
+        if cache is None:
+            cache = [None] * len(self.layers)
+
+        for e, layer in enumerate(self.layers):
+            h, cache[e] = layer(h, mask, cache[e])
+
+        return self.norm(h), cache
+
+class Model(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.args = args
+        self.model_type = args.model_type
+        self.model = MiniCPMModel(args)
+        self.lm_head = nn.Linear(args.hidden_size, args.vocab_size, bias=False)
+
+    def __call__(
+        self,
+        inputs: mx.array,
+        cache=None,
+    ):
+        out, cache = self.model(inputs, cache)
+        # out = out @ self.model.embed_tokens.weight.T
+        out = self.lm_head(out / (self.args.hidden_size / self.args.dim_model_base))
+        return out, cache
+
+    def sanitize(self, weights):
+        if "lm_head.weight" not in weights:
+            weights["lm_head.weight"] = weights["model.embed_tokens.weight"]
+        # Remove unused precomputed rotary freqs
+        return {
+            k: v for k, v in weights.items() if "self_attn.rotary_emb.inv_freq" not in k
+        }
```

Comment:
Do you need this sanitize method? It's better not to have it if you don't need.

Reply:
yes, sorry for the unnecessary code blocks, they can be removed.

Reply:
unfortunately when training without the sanitize method, I get the same error we had when training the Qwen2 models.

File "/Users/gokdenizgulmez/Desktop/1/mlx-examples/llms/mlx_lm/utils.py", line 349, in load_model
    model.load_weights(list(weights.items()))
                            ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'items'

so from my understanding, the sanitize method is needed for weight management.

Reply:
That error message doesn't make sense. I don't know why you are receiving it or why the `sanitize` removes it but the `sanitize` is not likely the right answer. We should debug what's actually causing that.

Reply:
I think I found out, what's causing it. Upon reviewing the actual model structure (`katuni4ka/tiny-random-minicpm`), I found that lm_head.weight is indeed represented by model.embed_tokens.weight. I have removed the rope code parts in the sanitized method and it works fine now. That error is still very weird and I have never gotten it after too. I think that that was a layer 8 error lol. When removing the sanitize method and running it, I get `ValueError: Missing parameters: lm_head.weight.`, which is a more understandable error.

```python
def sanitize(self, weights):
        if "lm_head.weight" not in weights:
            weights["lm_head.weight"] = weights["model.embed_tokens.weight"]
        return weights
 ```
 
So the sanitize method is to ensure that the lm_head.weight key is present and correctly assigned, especially when it's not explicitly set during model initialization.

That's at least from my understanding.

Reply:
Great! So rather than setting the `lm_head` to the embedding weight, it would be better to use it as a linear layer directly. That way parameters don't get duplicated taking up more memory.

I would do the following:
1. Remove `lm_head` from the model
2. Do `out @ model.embed_tokens.weight.T` to produce the final output.

Sound good?

Reply:
sounds great!

The new Model class is then:
```python
class Model(nn.Module):
    def __init__(self, args: ModelArgs):
        super().__init__()
        self.args = args
        self.model_type = args.model_type
        self.model = MiniCPMModel(args)

    def __call__(
        self,
        inputs: mx.array,
        cache=None,
    ):
        out, cache = self.model(inputs, cache)
        out = out @ self.model.embed_tokens.weight.T
        return out, cache
 ```
 
 
 Also training and inference is way better now too. Thanks for the quick response.

Reply:
Well when using the original model I get this error:

```
File "/Users/gokdenizgulmez/Desktop/1/mlx-examples/llms/mlx_lm/utils.py", line 349, in load_model
    model.load_weights(list(weights.items()))
  File "/Users/gokdenizgulmez/miniconda3/envs/mlx/lib/python3.12/site-packages/mlx/nn/layers/base.py", line 211, in load_weights
    raise ValueError(f"Received parameters not in model: {extras}.")
ValueError: Received parameters not in model: lm_head.weight.
```

Adding the lm_head part and the rest, trains the model right. It looks like the previous fix worked for `https://huggingface.co/katuni4ka/tiny-random-minicpm` but not for the original:  `https://huggingface.co/openbmb/MiniCPM-2B-128k/tree/aff9444f5c3f1795c2baddadfcf7be54d397ca11`

Reply:
Yup in that case you'll need to read the flag in the config `tie_word_embeddings` https://huggingface.co/openbmb/MiniCPM-2B-128k/blob/aff9444f5c3f1795c2baddadfcf7be54d397ca11/config.json#L33

If it is `True` (and maybe unset) then you would do as you have now. If it is `False` add separate `lm_head`

Reply:
Thank you for pointing me in the right direction regarding the tie_word_embeddings configuration. I have added the conditional handling of lm_head based on the flag, and the models are now performing as expected.